### PR TITLE
New smart contract to batch escrow transactions

### PR DESCRIPTION
**Pull request by avclarke**
Enables us to escrow batches of addresses and tokens at a single time to speed up distribution.

### CODE REVIEW

File: contracts/EscrowDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * @notice Set escrow amounts in batches.
+     * @dev required to approve this contract address to spend senders tokens before calling
+     * @param accounts: list of accounts to escrow
+     * @param amounts: corresponding list of amounts to escrow
+     * @param durationWeeks: number of weeks to escrow
+     */
+    function distirbuteEscrowed(
```
Comment (JChiaramonte7) [inline contracts/EscrowDistributor.sol:33]: Spelling

File: contracts/EscrowDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        address[] calldata accounts,
+        uint256[] calldata amounts,
+        uint256 durationWeeks
+    ) external {
+        require(
+            accounts.length == amounts.length,
+            "Number of accounts does not match number of values"
+        );
+
+        uint256 totalTokens = 0;
```
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:43]: default value is zero. does not need to be explicitly set to zero

File: contracts/EscrowDistributor.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        uint256 duration = durationWeeks * 1 weeks;
+
+        for (uint16 index = 0; index < accounts.length; index++) {
```
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:46]: can be optimized. See an example here: https://github.com/Kwenta/cc-vesting/blob/1f420f3fe7f342149214a488278cfca775eccaef/src/VestingFactory.sol#L129

Does not have to be a do-while but decrementing is cheaper and can be unchecked

File: contracts/EscrowDistributor.sol [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            IRewardEscrow(rewardEscrowAddr).createEscrowEntry(
+                accounts[index],
+                amounts[index],
+                duration
+            );
+        }
+
+        emit BatchEscrowed(totalTokens, accounts.length, durationWeeks);
```
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:61]: event params are not correct. Recommend the Event({paramName: name, (...) }) format

File: contracts/EscrowDistributor.sol [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        uint256 duration = durationWeeks * 1 weeks;
+
+        for (uint16 index = 0; index < accounts.length; index++) {
+            totalTokens += amounts[index];
+        }
+        IERC20 kwenta = IERC20(kwentaAddr);
+        kwenta.transferFrom(msg.sender, address(this), totalTokens);
+        kwenta.approve(rewardEscrowAddr, totalTokens);
+
+        for (uint16 index = 0; index < accounts.length; index++) {
```
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:53]: can optimize (see above)

File: test/local/contracts/EscrowDistributor.test.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        expect(
+            await distributor
+                .connect(owner)
+                .distirbuteEscrowed(
+                    [addr0.address, addr1.address, addr2.address],
+                    [wei(1).toBN(), wei(50).toBN(), wei(200).toBN()],
+                    DURATION_WEEKS
+                )
+        ).to.emit(distributor, "BatchEscrowed");
```
Comment (JaredBorders) [inline test/local/contracts/EscrowDistributor.test.ts:111]: Would be good to add test which ensure event params are correct

File: test/local/contracts/EscrowDistributor.test.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            wei(1).toBN()
+        );
+        expect(await rewardEscrow.balanceOf(addr1.address)).to.equal(
+            wei(50).toBN()
+        );
+        expect(await rewardEscrow.balanceOf(addr2.address)).to.equal(
+            wei(200).toBN()
+        );
+    });
+});
```
Comment (JaredBorders) [inline test/local/contracts/EscrowDistributor.test.ts:123]: would add tests for edge cases (i.e. when account address is zero or value is zero what happens?)

**Review by JaredBorders — CHANGES_REQUESTED** [review]
Left several comments! lmk if you have any questions.

File: contracts/EscrowDistributor.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "./interfaces/IRewardEscrow.sol";
+
+contract EscrowDistributor {
+    /// @notice kwenta token contract
+    address public immutable kwentaAddr;
```
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:10]: if you save this as IERC20 vs address, would that save gas? Would be good to know. That way you don't have to do the `IERC20(kwenta).transferFrom()` and can just do it `kwenta.transferFrom()`.

May not make a difference

File: test/local/contracts/EscrowDistributor.test.ts (hunk 6 above)
Comment (avclarke) [reply test/local/contracts/EscrowDistributor.test.ts:111]: is there a different function for this, it seems `emit` only takes contract and eventName?

Comment (JaredBorders) [reply test/local/contracts/EscrowDistributor.test.ts:111]: Yes there is. Here is an example I have used in the past but there is also an easier way that is a little less verbose

https://github.com/Kwenta/margin-manager/blob/906267e512f4242274235800f3e03370b04d0e7d/test/integration/margin.behavior.ts#L1420

File: contracts/EscrowDistributor.sol (hunk 8 above)
Comment (avclarke) [reply contracts/EscrowDistributor.sol:10]: Yea I did have it like that and then thought it might be handy to store address if ever anyone wanted to read which token address it was, but probably not really needed. Is there a simple way you compare gas, can you get that gas value when calling in a test?

File: test/local/contracts/EscrowDistributor.test.ts (hunk 6 above)
Comment (JChiaramonte7) [reply test/local/contracts/EscrowDistributor.test.ts:111]: You can chain `emit()` with `withArgs(arg1, arg2, arg3...)`

https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#emitting-events

File: contracts/EscrowDistributor.sol (hunk 8 above)
Comment (JChiaramonte7) [reply contracts/EscrowDistributor.sol:10]: I believe even if you do store an address as a contract type, it will still return the address offchain (ie. on etherscan)

Example of StakingRewards if you try to read rewardEscrow or supplySchedule: https://goerli-optimism.etherscan.io/address/0x1653a3a3c4ccee0538685f1600a30df5e3ee830a#readContract

Comment (avclarke) [reply contracts/EscrowDistributor.sol:10]: ok good to know thanks, now storing both rewardEscrow and kwenta as contract instances instead of addresses

File: test/local/contracts/EscrowDistributor.test.ts (hunk 6 above)
Comment (avclarke) [reply test/local/contracts/EscrowDistributor.test.ts:111]: perfect, thanks

File: contracts/EscrowDistributor.sol (hunk 8 above)
Comment (JaredBorders) [reply contracts/EscrowDistributor.sol:10]: So preference I suppose. And as far as testing gas costs, hardhat has a plugin. Foundry reports it by default which is what i was using to test out my theory haha

**Review by JaredBorders — APPROVED** [review]
Looks good to me. I'd wait for Jeremy to review too before merging.

Nice work!!

File: contracts/EscrowDistributor.sol (hunk 8 above)
Comment (avclarke) [reply contracts/EscrowDistributor.sol]: Good to know

**Review by JChiaramonte7 — APPROVED** [review]
